src/ucs/stats/stats.c

```diff
@@ -460,6 +481,197 @@ void ucs_stats_node_free(ucs_stats_node_t *node)
+    /* Initialize return value */
+    offset = ucs_stats_context.aggrgt_sum_db_last_offset;
+
+    /* Return class_id */
+    node->cls->class_id = db_size;
+
+    /* Update global offset */
+    ucs_stats_context.aggrgt_sum_db_last_offset += n_cnts_in_cls;
+
+    /* Reallocate counters names db */
+    db_size = ucs_stats_context.aggrgt_cnt_names_db_size += n_cnts_in_cls;
+
+    total_alloc_size_bytes =
+            (db_size * sizeof(*ucs_stats_context.aggrgt_cnt_names_db));
+    ucs_stats_context.aggrgt_cnt_names_db = ucs_realloc(
+            ucs_stats_context.aggrgt_cnt_names_db, total_alloc_size_bytes,
+            (const char*)"stats_cnt_name_db");
+
+    cnt_index = offset;
+    ucs_for_each_bit(counter_index, filter_node->counters_bitmask) {
+        ucs_list_for_each(temp_node, &filter_node->type_list_head,
+                          type_list) {
+            ucs_stats_context.aggrgt_cnt_names_db[cnt_index].name =
+                    node->cls->counter_names[counter_index];
+            ucs_stats_context.aggrgt_cnt_names_db[cnt_index].class_name =
+                    node->cls->name;
+        }
+        cnt_index++;
+    }
+
+    ucs_stats_context.aggrgt_sum_db_size++;
+
+    return offset;
+}
+
+static uint32_t
+ucs_stats_clsid_to_base_offset(ucs_stats_node_t *node,
+                                             size_t num_counters_in_class)
+{
+    uint32_t offset = 0;
+    ucs_stats_aggrgt_sum_clsid_to_name_db_t *element;
+
+    pthread_mutex_lock(&ucs_stats_context.lock);
+
+    if (ucs_likely(node->cls->class_id != 0)) {
+        element = &ucs_array_elem(&ucs_stats_context.aggrgt_sum_db,
+                                  node->cls->class_id);
+        offset = element->aggregate_sum_class_offset;
+
+        goto offset_get_exit;
+    } else {
+        offset = ucs_stats_agrgt_sum_clsid_alloc(
+                         node, num_counters_in_class);
+    }
+
+offset_get_exit:
+    pthread_mutex_unlock(&ucs_stats_context.lock);
+
+    return offset;
+}
+
+
+/**
+ * Returns the aggregate_sum of all counters from all interfaces for a compact statistics trace.
+ * Used for Score-P UCX plugin to collect a compact trace of the counters.
+ *
+ * @param [in]  node                      Pointer to the statistics node.
+ * @param [in]  sel                       Whether or not to recursively iterate through the children.
+ * @param [in]  current_num_counters      Current number of counters iterated through (call with 0).
+ * @param [out] ucs_stats_aggregate_sum   The aggregate-sum of all counters of each class/type.
+ * @param [in]  num_stats_counters_max The number of elements in ucs_stats_aggregate_sum.
+ *
+ * Complexity: O(n)
+ *
+ * Returns: The total number of counters.
+ *
+ * Remarks: The function is recursive.
+ */
+static size_t ucs_stats_aggregate_sum_recurs(ucs_stats_node_t *node,
+                                             ucs_stats_children_sel_t sel,
+                                             size_t current_num_counters,
```

Comment:
can we use aggrgt_sum_db_last_offset instead of passing this parameter?

Reply:
They are different parameters. n_counters is the total number of counters seen on the list, while aggrgt_sum_db_last_offset is the total number of aggregate-sum counters (a much smaller number, a unique counter is stored per class/type of counter).

Reply:
why do we care about total number of counters? seems it's neither used anywhere and also not returned from API